tests.test_listing.test_directory fails on Debian/Ubuntu systems (`GDAL_DATA` not found)

Fiona checks for common `GDAL_DATA` paths in `fiona/_drivers.pyx`, but not for the one used by the Debian package:

`/usr/share/gdal/<GDAL_VERSION_MAJOR>.<GDAL_VERSION_MINOR>`

This causes some `tests.test_listing` tests to fail:
```
======================================================================
FAIL: tests.test_listing.test_directory
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/<<PKGBUILDDIR>>/.pybuild/cpython2_2.7_fiona/build/tests/test_listing.py", line 22, in test_directory
    assert fiona.listlayers('tests/data') == ['coutwildrnp', 'gre']
AssertionError: 
-------------------- >> begin captured logging << --------------------
fiona: DEBUG: Creating a not-responsible GDALEnv in drivers()
Fiona: WARNING: GDAL data files not located, GDAL_DATA not set
Fiona: DEBUG: PROJ_LIB: /usr/share/proj
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: tests.test_listing.test_directory_trailing_slash
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/<<PKGBUILDDIR>>/.pybuild/cpython2_2.7_fiona/build/tests/test_listing.py", line 27, in test_directory_trailing_slash
    assert fiona.listlayers('tests/data/') == ['coutwildrnp', 'gre']
AssertionError: 
-------------------- >> begin captured logging << --------------------
fiona: DEBUG: Creating a not-responsible GDALEnv in drivers()
Fiona: WARNING: GDAL data files not located, GDAL_DATA not set
Fiona: DEBUG: PROJ_LIB: /usr/share/proj
--------------------- >> end captured logging << ---------------------
```

## Steps to reproduce the problem.

Run tests on a Debian/Ubuntu system with the packaged gdal (libgdal-dev)

## Operating system

For example: Debian unstable

## Fiona and GDAL version and provenance

Fiona 1.7.11 (Debian source package)

GDAL 2.2.4 (Debian package installed as part of the fiona build dependencies).

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -31,6 +31,8 @@ Bug fixes:
 - Revert the change in 0a2bc7c that discards Z in geometry types when a
   collection's schema is reported (#541).
 - Require six version 1.7 or higher (#550).
+- A regression related to "zip+s3" URIs has been fixed.
+- Debian's GDAL data locations are now searched by default (#583).
 
 1.8a1 (2017-11-06)
 ------------------

--- a/fiona/_drivers.pyx
+++ b/fiona/_drivers.pyx
@@ -28,6 +28,7 @@ cdef extern from "gdal.h":
     void * GDALGetDriver(int i)
     const char * GDALGetDriverShortName(void *driver)
     const char * GDALGetDriverLongName(void *driver)
+    const char * GDALVersionInfo(const char *key)
 
 
 cdef extern from "ogr_api.h":
@@ -111,15 +112,33 @@ cdef class GDALEnv(object):
         if 'GDAL_DATA' in os.environ:
             log.debug("GDAL_DATA: %s", os.environ['GDAL_DATA'])
         else:
+            # We will try a few well-known paths, starting with the
+            # official wheel path.
             whl_datadir = os.path.abspath(
                 os.path.join(os.path.dirname(__file__), "gdal_data"))
-            share_datadir = os.path.join(sys.prefix, 'share/gdal')
+            fhs_share_datadir = os.path.join(sys.prefix, 'share/gdal')
+
+            # Debian supports multiple GDAL installs.
+            gdal_release_name = GDALVersionInfo("RELEASE_NAME")
+            gdal_release_name = gdal_release_name.decode('utf-8')
+            deb_share_datadir = os.path.join(
+                fhs_share_datadir,
+                "{}.{}".format(*gdal_release_name.split('.')[:2]))
+
+            # If we find GDAL data at the well-known paths, we will
+            # add a GDAL_DATA key to the config options dict.
             if os.path.exists(os.path.join(whl_datadir, 'pcs.csv')):
-                os.environ['GDAL_DATA'] = whl_datadir
                 log.debug("Set GDAL_DATA = %r", whl_datadir)
-            elif os.path.exists(os.path.join(share_datadir, 'pcs.csv')):
-                os.environ['GDAL_DATA'] = share_datadir
-                log.debug("Set GDAL_DATA = %r", share_datadir)
+                self.options['GDAL_DATA'] = whl_datadir
+
+            elif os.path.exists(os.path.join(deb_share_datadir, 'pcs.csv')):
+                log.debug("Set GDAL_DATA = %r", deb_share_datadir)
+                self.options['GDAL_DATA'] = deb_share_datadir
+
+            elif os.path.exists(os.path.join(fhs_share_datadir, 'pcs.csv')):
+                os.environ['GDAL_DATA'] = fhs_share_datadir
+                log.debug("Set GDAL_DATA = %r", fhs_share_datadir)
+
             else:
                 log.warning("GDAL data files not located, GDAL_DATA not set")
 
@@ -129,12 +148,15 @@ cdef class GDALEnv(object):
             whl_datadir = os.path.abspath(
                 os.path.join(os.path.dirname(__file__), "proj_data"))
             share_datadir = os.path.join(sys.prefix, 'share/proj')
+
             if os.path.exists(whl_datadir):
-                os.environ['PROJ_LIB'] = whl_datadir
                 log.debug("Set PROJ_LIB = %r", whl_datadir)
+                self.options['PROJ_LIB'] = whl_datadir
+
             elif os.path.exists(share_datadir):
-                os.environ['PROJ_LIB'] = share_datadir
                 log.debug("Set PROJ_LIB = %r", share_datadir)
+                self.options['PROJ_LIB'] = share_datadir
+
             else:
                 log.warning("PROJ data files not located, PROJ_LIB not set")
 

